File: Data Structures & Algorithms/combination-target-sum/submission-0.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        """
        nums = [2 5 6 9]
        target = 9

        with reuse
        """

        path: List[int] = []
        result: List[List[int]] = []

        def findPath(nums: List[int], idx: int, target: int):
            if target < 0 or idx >= len(nums):
                return

            if target == 0:
                result.append(path[:])
                return
            
            
            path.append(nums[idx])
            findPath(nums, idx, target - nums[idx])
            path.pop()

            findPath(nums, idx+1, target)
            
        findPath(nums,0,target)
        return result 

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int start = 0;
        int end = n-1;
        while(start <= end){
            int mid = (start + end)/2;
            if(nums[mid] == target){
                return mid;
            }

            if(nums[mid] >= nums[start]){
                if(nums[mid] >= target && target >= nums[start]){
                    end = mid - 1;
                }
                else {
                    start = mid + 1;
                }
            }
            else {
                if(nums[end] >= target && target >= nums[mid]){
                    start = mid + 1;
                }
                else{
                    end = mid - 1;
                }
            }
        }
        return -1;
    }
};
